### ISSUE TITLE
When implementing a custom value generator, it's not clear how to implement async version

### ISSUE THREAD

**Issue by obohaciak** [dotnet/efcore#20671]
When implementing a custom value generator based on `ValueGenerator<T>`, it's not obvious how to implement an _**async**_ value generation. 

Sync value generation is straightforward; one only needs to (in fact, has to) override 

`public new abstract TValue Next(EntityEntry entry)`

One would think that the analogous method 

`public new virtual ValueTask<TValue> NextAsync(EntityEntry entry, CancellationToken cancellationToken)` 

is the way to do it. That is not the case, as there's a broken call chain due to the **`new`** modifier on the `NextAsync` method. Due to this modifier, the overridden `NextAsync` method never gets called; instead it's the base non-generic `ValueGenerator`'s `NextAsync` that gets called which routes the call to synchronous `Next`. This broken call chain was already reported in [issue 12900](https://github.com/dotnet/efcore/issues/12900#issuecomment-411389952) by @gilmishal

### Steps to reproduce

Let's suppose I need to generate permanent IDs in the DB using a stored procedure. While the stored procedure executes, I'd like to free up the thread and let it work on something else until I/O finishes. I provide my own value generator that calls the stored procedure, provide both sync & async versions and configure it in the model. See sample below:

```csharp
internal class IdValueGenerator : ValueGenerator<int>
{
    public override bool GeneratesTemporaryValues => false;

    // sync override - called by DbContext.Add()
    public override int Next(EntityEntry entry)
        => CallStoredProcedureAsync(entry).AsTask().GetAwaiter().GetResult();

    // never called!
    public override ValueTask<int> NextAsync(EntityEntry entry, CancellationToken cancellationToken = default)
        => CallStoredProcedureAsync(entry);

    protected override object NextValue(EntityEntry entry)
        => CallStoredProcedureAsync(entry).AsTask().GetAwaiter().GetResult();

    // async override - called by DbContext.AddAsync()
    protected override async ValueTask<object> NextValueAsync(EntityEntry entry, CancellationToken cancellationToken = default)
        => CallStoredProcedureAsync(entry);
}
```

Now when I add a new entity to the `DbContext` and expect it to call `NextAsync`, what happens instead is that `NextValueAsync` gets called.

```
Customer customer = NewCustomer();
await db.AddAsync(customer);
```

### TL;DR

My gripe with this is that as a developer I don't know which method should I override for _**async**_ value generation without reading through EF Core source code. There's the more obvious public one (`NextAsync`) that matches the generic type parameter and has a name that suggests that could be it or the protected one (`NextValueAsync`) that seems to be there as internal implementation detail? As it turns out, it's the latter, which I find surprising.

Can we get this either fixed or documented better? 

I suppose a simple fix would be in `ValueGenerator<T>` to have `NexValueAsync` call `NextAsync` instead of wrapping a call to `Next` in a `ValueTask`.

A documentation that would point me to the right method would be also welcome.

### Further technical details

EF Core version:
Database provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer)
Target framework: (e.g. .NET Core 3.1.2)
Operating system: Windows 10
IDE: (e.g. Visual Studio 2019 16.5.3)

**Comment by ajcvickers** [comment]
@obohaciak This is mostly a duplicate of #19420. I will add a note there to make sure the documentation is helpful and correct as well.
